### PR TITLE
Reviewed the Bundles cookbook articles

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Doc fix? | yes |
| New docs? | no |
| Applies to | all |
| Fixed tickets | - |

### CODE REVIEW

File: cookbook/bundles/installation.rst [hunk 1]
```diff
@@ -68,22 +71,46 @@ The only thing you need to do now is register the bundle in ``AppKernel``::
         }
     }
 
+By default, Symfony bundles are registered in all the application
+:doc:`execution environments <cookbook/configuration/environments>`. If the bundle
```
Comment (xabbuh) [inline cookbook/bundles/installation.rst]: missing the leading slash here

Comment (javiereguiluz) [inline cookbook/bundles/installation.rst]: Fixed. Thanks.

File: cookbook/bundles/best_practices.rst [hunk 2]
```diff
@@ -190,13 +186,14 @@ the ``Tests/`` directory. Tests should follow the following principles:
...
 .. note::
+
    A test suite must not contain ``AllTests.php`` scripts, but must rely on the
    existence of a ``phpunit.xml.dist`` file.
 
 Documentation
 -------------
 
-All classes and functions must come with full PHPDoc.
+All classes and functions must be fully documented using `phpDocumentor`_ tags.
```
Comment (WouterJ) [inline cookbook/bundles/best_practices.rst]: why phpDocumentor if Symfony relies on Sami and other PHPdoc generators are popular too?

File: cookbook/bundles/best_practices.rst [hunk 3]
```diff
@@ -221,7 +218,7 @@ following standardized instructions in your ``README.md`` file.
     following command to download the latest stable version of this bundle:
 
     ```bash
-    $ composer require <package-name> "~1"
+    $ composer require <package-name>
```
Comment (WouterJ) [inline cookbook/bundles/best_practices.rst]: please, do not remove this version number. If you have documentation for 1.0 version of your docs, it should explain installing version 1. If you release version 2.0 of your library, the 1.0 version of your docs should still explain the version 1 installation. Removing the version number means the 1.0 version of the docs makes you install 2.0

File: cookbook/bundles/extension.rst [hunk 4]
```diff
@@ -15,7 +15,7 @@ Creating an Extension Class
 ---------------------------
 
 In order to load service configuration, you have to create a Dependency
-Injection Extension for your bundle. This class has some conventions in order
+Injection (DI) Extension for your bundle. This class has some conventions in order
```
Comment (WouterJ) [inline cookbook/bundles/extension.rst]: I don't really get why this change is usefull

File: cookbook/bundles/installation.rst [hunk 5]
```diff
@@ -39,9 +39,12 @@ Now that you know the package name, you can install it via Composer:
...
 This will choose the best version for your project, add it to ``composer.json``
-and download the library into the ``vendor/`` directory. If you need a specific
-version, add a ``:`` and the version right after the library name (see
-`composer require`_).
+and download its code into the ``vendor/`` directory. If you need a specific
+version, include it as the second argument of the `composer require`_ command:
+
+.. code-block:: bash
+
+    $ composer require friendsofsymfony/user-bundle "~2.0@dev"
```
Comment (WouterJ) [inline cookbook/bundles/installation.rst]: I'm not sure showing a dev version is great and not confusing.

File: cookbook/bundles/installation.rst [hunk 6]
```diff
@@ -68,22 +71,46 @@ The only thing you need to do now is register the bundle in ``AppKernel``::
         }
     }
 
+By default, Symfony bundles are registered in all the application
+:doc:`execution environments </cookbook/configuration/environments>`. If the bundle
+is meant to be used only in the development or test environments, register it in
+the section below::
```
Comment (WouterJ) [inline cookbook/bundles/installation.rst]: I would replace this sentence by "If the bundle is meant to be used only in some environment, register it within an if statement, like this (where the FOSUserBundle is only available in the dev and test environment)::"

File: cookbook/bundles/installation.rst [hunk 7]
```diff
@@ -68,22 +71,46 @@ The only thing you need to do now is register the bundle in ``AppKernel``::
...
+
+    // ...
+    class AppKernel extends Kernel
+    {
+        // ...
+
+        public function registerBundles()
+        {
+            $bundles = array(
+                // ...,
```
Comment (WouterJ) [inline cookbook/bundles/installation.rst]: we decided to remove the comma

File: cookbook/bundles/extension.rst (hunk 4 above)
Comment (javiereguiluz) [inline cookbook/bundles/extension.rst]: A good practice when using acronyms is to introduce them properly before the first time they are used it. In this article, we talk about `Dependency Injection` and suddenly, the article talks about `DI`. I know that Symfony developers are smart enough to infere that `DI` means `Dependency Injection`, but the usual good practice is to show the full words + its acronym just before starting to use the acronym exclusively.

File: cookbook/bundles/installation.rst (hunk 5 above)
Comment (javiereguiluz) [inline cookbook/bundles/installation.rst]: This is just an example of using specific version constraints, so I think it doesn't matter to use a dev version. In fact, the example is real because it was copied from the FOSUserBundle documentation.

File: cookbook/bundles/best_practices.rst (hunk 3 above)
Comment (javiereguiluz) [inline cookbook/bundles/best_practices.rst]: I agree with you ... but at the same time, for DX reasons, we don't want to promote the use of versions in all cases. I propose to add a note explaining when do you need to add the version number. See 29fe426.

Comment (WouterJ) [inline cookbook/bundles/best_practices.rst]: Is typing 2 more characters already seen as a decrease in DX? I think adding a 5 line long paragraph about a simple version number is not needed and makes the docs harder to read.

And indeed, we shouldn't promote it to use everywhere, but it should be used in the installation instructions. A case where a version isn't needed is: "This bundle can also work with xxx/yyy. You can install it using `composer require xxx/yyy`". However, this section isn't about that case.

File: cookbook/bundles/installation.rst (hunk 5 above)
Comment (WouterJ) [inline cookbook/bundles/installation.rst]: It is a real example, but it will promote to install dev versions (not all people understand composer version constraints as good as we do).

**Comment by javiereguiluz** [comment]
I've just fixed the conflicts of this PR. I think it's ready to be merged, because Wouter's concerns are taken into account (the first example removes the version number because of how we explain things and there is a note explaining that ... the second example shows a `@dev` version as an example of using advanced version constraints, so it's not unsafe to do that).

File: cookbook/bundles/best_practices.rst (hunk 3 above)
Comment (xabbuh) [inline cookbook/bundles/best_practices.rst]: I would leave the version number. If you don't add it, you will very likely get into support trouble one you release a new major version and people complain about BC breaks because of reading old documentations.

File: cookbook/bundles/extension.rst (hunk 4 above)
Comment (xabbuh) [inline cookbook/bundles/extension.rst]: We could also just avoid using DI in the docs (which imho were a better practice). But as long as we use them it's imho indeed a good idea to introduce it properly.

File: cookbook/bundles/installation.rst (hunk 5 above)
Comment (xabbuh) [inline cookbook/bundles/installation.rst]: Taking a real-world example is a good idea. But can't we find a better example (without the need to use a dev version)?

File: cookbook/bundles/best_practices.rst (hunk 3 above)
Comment (javiereguiluz) [inline cookbook/bundles/best_practices.rst]: OK, I've added the version number back.

File: cookbook/bundles/installation.rst (hunk 5 above)
Comment (javiereguiluz) [inline cookbook/bundles/installation.rst]: OK, I've removed the `@dev` suffix.

**Comment by javiereguiluz** [comment]
I've made all the proposed changes and I've fixed all the conflicts. Please, review it for the last time, label it as `finished` and consider merging it. Thanks.

**Comment by xabbuh** [comment]
:+1:

**Comment by weaverryan** [comment]
Thanks Javier! Minor tweaks at sha: ae3533813f02ab6c8e73ccc643269b2de93f5ec4
